tests/xharness/Jenkins/TestSelector.cs

```diff
@@ -202,67 +227,53 @@ void SelectTestsByLabel (int pullRequest)
-			SetEnabled (labels, "monotouch", ref jenkins.IncludeMonotouch);
+			foreach (var keyValuePair in  disabledByDefault) {
+				SetEnabled (labels, keyValuePair.Key, keyValuePair.Value);
+			}
+			
+			foreach (var keyValuePair in enabledByDefault) {
+				SetEnabled (labels, keyValuePair.Key, keyValuePair.Value);
+			}
 
-			bool inc_permission_tests = false;
-			if (SetEnabled (labels, "system-permission", ref inc_permission_tests))
-				Harness.IncludeSystemPermissionTests = inc_permission_tests;
+			// special case since we are not working with the TestSelection flag in jenkins
+			Harness.IncludeSystemPermissionTests = labels.Contains ("run-system-permission-tests");
 
 			// docs is a bit special:
 			// - can only be executed if the Xamarin-specific parts of the build is enabled
 			// - enabled by default if the current branch is master (or, for a pull request, if the target branch is master)
-			var changed = SetEnabled (labels, "docs", ref jenkins.IncludeDocs);
+			var changed = SetEnabled (labels, "docs", TestSelection.Docs);
 			if (Harness.ENABLE_XAMARIN) {
 				if (!changed) { // don't override any value set using labels
 					var branchName = Environment.GetEnvironmentVariable ("BRANCH_NAME");
 					if (!string.IsNullOrEmpty (branchName)) {
-						jenkins.IncludeDocs = branchName == "master";
-						if (jenkins.IncludeDocs)
+						if (branchName == "master") {
+							jenkins.TestSelection |= TestSelection.Docs;
+						} else {
+							jenkins.TestSelection &= ~TestSelection.Docs;
+						}
+						if (jenkins.TestSelection.HasFlag (TestSelection.Docs))
 							MainLog.WriteLine ("Enabled 'docs' tests because the current branch is 'master'.");
 					} else if (pullRequest > 0) {
-						jenkins.IncludeDocs = vcs.GetPullRequestTargetBranch (pullRequest) == "master";
-						if (jenkins.IncludeDocs)
+						if (vcs.GetPullRequestTargetBranch (pullRequest) == "master") {
+							jenkins.TestSelection |= TestSelection.Docs;
+						} else {
+							jenkins.TestSelection &= ~TestSelection.Docs;
+						}
+						if (jenkins.TestSelection.HasFlag (TestSelection.Docs))
 							MainLog.WriteLine ("Enabled 'docs' tests because the target branch is 'master'.");
 					}
 				}
 			} else {
-				if (jenkins.IncludeDocs) {
-					jenkins.IncludeDocs = false; // could have been enabled by 'run-all-tests', so disable it if we can't run it.
+				if (jenkins.TestSelection.HasFlag (TestSelection.Docs)) {
+					jenkins.TestSelection &= ~TestSelection.Docs; // could have been enabled by 'run-all-tests', so disable it if we can't run it.
 					MainLog.WriteLine ("Disabled 'docs' tests because the Xamarin-specific parts of the build are not enabled.");
 				}
 			}
 
 			// old simulator tests is also a bit special:
 			// - enabled by default if using a beta Xcode, otherwise disabled by default
-			changed = SetEnabled (labels, "old-simulator", ref jenkins.IncludeOldSimulatorTests);
 			if (!changed && jenkins.IsBetaXcode) {
```

Comment:
You're still checking `changed` even if you removed the previous line that set `changed`.